clang/runtime/dpct-rt/include/dpct/math.hpp

```diff
@@ -2216,6 +2218,103 @@ void ldmatrix(uintptr_t addr, T *m1, T *m2, T *m3, T *m4, bool trans = false) {
+/// \param [in] c_mat The elements of the input C matrix to be added with the
+/// result of A * B
+template <int M, int N, int K, typename MulType, typename ABType,
+          typename CDType>
+void mma(void **d_mat, void *a_mat, void *b_mat, void *c_mat) {
+  auto d = reinterpret_cast<CDType **>(d_mat);
+  auto a = reinterpret_cast<ABType *>(a_mat);
+  auto b = reinterpret_cast<ABType *>(b_mat);
+  auto c = reinterpret_cast<CDType *>(c_mat);
+
+  auto sg = sycl::ext::oneapi::this_work_item::get_sub_group();
+  int lane = sg.get_local_linear_id();
+
+  static_assert(M == 16 && N == 8 && K == 16,
+                "Only m16n8k16 shape is supported!");
+
+  short ROW_LOAD_OFFSET = 4 * (lane >> 2);
+  short COL_LOAD_OFFSET = 8 * (lane % 4);
+
+  if constexpr (M == 16 && N == 8 && K == 16) {
+    if constexpr (std::is_floating_point_v<CDType>) {
+      for (int i = 0; i < 4; i++) {
+        ABType recv_a[4], recv_b[4];
+
+        recv_a[0] = dpct::select_from_sub_group(sg, a[0], ROW_LOAD_OFFSET + i);
+        recv_a[1] = dpct::select_from_sub_group(sg, a[2], ROW_LOAD_OFFSET + i);
+        recv_a[2] = dpct::select_from_sub_group(sg, a[1], ROW_LOAD_OFFSET + i);
+        recv_a[3] = dpct::select_from_sub_group(sg, a[3], ROW_LOAD_OFFSET + i);
+
+        recv_b[0] = dpct::select_from_sub_group(sg, b[0], COL_LOAD_OFFSET + i);
+        recv_b[1] = dpct::select_from_sub_group(sg, b[1], COL_LOAD_OFFSET + i);
+        recv_b[2] =
+            dpct::select_from_sub_group(sg, b[0], COL_LOAD_OFFSET + 4 + i);
+        recv_b[3] =
+            dpct::select_from_sub_group(sg, b[1], COL_LOAD_OFFSET + 4 + i);
+
+        auto ra = reinterpret_cast<MulType *>(recv_a);
+        auto rb = reinterpret_cast<MulType *>(recv_b);
+
+        for (int j = 0; j < 4; j++) {
+          c[0] += static_cast<CDType>(ra[j]) * static_cast<CDType>(rb[j]);
+          c[1] += static_cast<CDType>(ra[j]) * static_cast<CDType>(rb[j + 4]);
+          c[2] += static_cast<CDType>(ra[j + 4]) * static_cast<CDType>(rb[j]);
+          c[3] +=
+              static_cast<CDType>(ra[j + 4]) * static_cast<CDType>(rb[j + 4]);
+        }
+      }
+
+      *d[0] = c[0];
+      *d[1] = c[1];
+      *d[2] = c[2];
+      *d[3] = c[3];
+    } else if constexpr (std::is_integral_v<MulType>) {
+      for (int i = 0; i < 4; i++) {
+        ABType recv_a[2], recv_b[2];
+
+        recv_a[0] = dpct::select_from_sub_group(sg, a[0], ROW_LOAD_OFFSET + i);
+        recv_a[1] = dpct::select_from_sub_group(sg, a[1], ROW_LOAD_OFFSET + i);
+        recv_b[0] = dpct::select_from_sub_group(sg, b[0], COL_LOAD_OFFSET + i);
+        recv_b[1] =
+            dpct::select_from_sub_group(sg, b[0], COL_LOAD_OFFSET + i + 4);
+
+        auto ra = reinterpret_cast<MulType *>(recv_a);
+        auto rb = reinterpret_cast<MulType *>(recv_b);
+        for (int i = 0; i < 4; i++) {
+          c[0] += ra[i] * rb[i];
+          c[1] += ra[i] * rb[i + 4];
+          c[2] += ra[i + 4] * rb[i];
+          c[3] += ra[i + 4] * rb[i + 4];
```

Comment:
Here need to define a separate array[4] like "float c[4] = {0.0, 0.0, 0.0, 0.0};" to record the accumulated sum value if CDType is float instead of using `c` array.

Reply:
Retained the function signature to pass the value of 'c' to match PTX syntax